File: vulns/injection.js
// Intentionally vulnerable Express lab for SAST validation. DO NOT DEPLOY.
const express = require('express');
const child_process = require('child_process');
const vm = require('vm');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { MongoClient } = require('mongodb');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Hardcoded fake secrets with realistic shapes
// CWE-798: Use of Hard-coded Credentials
const AWS_SECRET_ACCESS_KEY = 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY';
const STRIPE_API_KEY = 'sk_live_51Hq8zKJ9aB2cD3eF4gH5iJ6kL7mN8oP9qR0sT1uV2wX3yZ';

let db; // SQL-like connection stub
const mongo = { db: null };

// CWE-95: Eval Injection — query param flows directly into eval()
app.get('/eval', (req, res) => {
  const x = req.query.x;
  const result = eval(x);
  res.send(String(result));
});

// CWE-78: OS Command Injection — query param into shell command
app.get('/ping', (req, res) => {
  const cmd = req.query.cmd;
  child_process.exec(`ping ${cmd}`, (err, stdout) => {
    res.send(stdout || String(err));
  });
});

// CWE-78: OS Command Injection — execSync variant from request body
app.post('/run', (req, res) => {
  const target = req.body.target;
  const out = child_process.execSync('nslookup ' + target);
  res.send(out.toString());
});

// CWE-94: Code Injection via vm.runInThisContext
app.post('/vm', (req, res) => {
  const code = req.body.code;
  const out = vm.runInThisContext(code);
  res.send(String(out));
});

// CWE-94: Code Injection via vm.runInNewContext
app.get('/sandbox', (req, res) => {
  const expr = req.query.expr;
  const out = vm.runInNewContext(expr, {});
  res.send(String(out));
});

// CWE-89: SQL Injection — string concatenation of request body
app.post('/user', (req, res) => {
  const id = req.body.id;
  db.query("SELECT * FROM users WHERE id = " + id, (err, rows) => {
    res.json(rows);
  });
});

// CWE-89: SQL Injection — template literal with query param
app.get('/search', (req, res) => {
  const name = req.query.name;
  db.query(`SELECT * FROM products WHERE name = '${name}'`, (err, rows) => {
    res.json(rows);
  });
});

// CWE-22: Path Traversal — unsanitized path read from request
app.get('/file', (req, res) => {
  const p = req.query.path;
  fs.readFile(p, 'utf8', (err, data) => {
    res.send(data || String(err));
  });
});

// CWE-22: Path Traversal — join with user-controlled filename
app.get('/download', (req, res) => {
  const name = req.query.name;
  const full = path.join('/var/data/', name);
  res.send(fs.readFileSync(full));
});

// CWE-98: Unsafe require() of user-controlled module name
app.get('/load', (req, res) => {
  const mod = req.query.module;
  const loaded = require(mod);
  res.send(Object.keys(loaded));
});

// CWE-943: NoSQL Injection — $where operator from request body
app.post('/find', async (req, res) => {
  const q = req.body.q;
  const docs = await mongo.db.collection('accounts').find({ $where: q }).toArray();
  res.json(docs);
});

// CWE-943: NoSQL Injection — raw object from query into filter
app.get('/lookup', async (req, res) => {
  const user = req.query.user;
  const docs = await mongo.db.collection('users').find({ username: user }).toArray();
  res.json(docs);
});

// CWE-1321: Prototype Pollution via recursive merge of request body
function merge(target, source) {
  for (const key in source) {
    if (typeof source[key] === 'object' && source[key] !== null) {
      if (!target[key]) target[key] = {};
      merge(target[key], source[key]);
    } else {
      target[key] = source[key];
    }
  }
  return target;
}
app.post('/config', (req, res) => {
  const settings = {};
  merge(settings, req.body);
  res.json(settings);
});

// CWE-918: SSRF — fetch a URL supplied by the user
app.get('/fetch', (req, res) => {
  const url = req.query.url;
  require('http').get(url, (r) => {
    let body = '';
    r.on('data', (c) => (body += c));
    r.on('end', () => res.send(body));
  });
});

// CWE-79: Reflected XSS — request input written into HTML response
app.get('/hello', (req, res) => {
  const name = req.query.name;
  res.send(`<html><body><h1>Hello ${name}</h1></body></html>`);
});

// CWE-601: Open Redirect — redirect to user-controlled location
app.get('/go', (req, res) => {
  const next = req.query.next;
  res.redirect(next);
});

// CWE-330: Weak token derived from predictable, attacker-influenced seed
app.get('/token', (req, res) => {
  const seed = req.query.seed;
  const token = crypto.createHash('md5').update(seed).digest('hex');
  res.send(token);
});

app.listen(3000, () => console.log('vuln lab on 3000'));

module.exports = { app, merge };

File: vulns/injection.py
"""Intentionally vulnerable injection lab (DVWA-style). For SAST validation ONLY.

Every handler reads untrusted input and flows it into a dangerous sink with no
sanitization. Do NOT deploy. Each function names the CWE it demonstrates.
"""
import os
import sys
import pickle
import sqlite3
import subprocess

import yaml
from flask import Flask, request
from fastapi import FastAPI, Query

app = Flask(__name__)
api = FastAPI()
db = sqlite3.connect(":memory:", check_same_thread=False)

# --- hardcoded fake secrets (CWE-798: Use of Hard-coded Credentials) ---
AWS_SECRET_ACCESS_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
STRIPE_API_KEY = "sk_live_51HxQ2eJ8sZ4kPq9mNvTbWxYz0aBcDeFgHiJkLmNoP"


@app.route("/sqli/fstring")
def sqli_fstring():
    # CWE-89: SQL Injection via f-string interpolation into execute()
    uid = request.args.get("id")
    cur = db.cursor()
    cur.execute(f"SELECT * FROM users WHERE id = {uid}")
    return str(cur.fetchall())


@app.route("/sqli/format")
def sqli_format():
    # CWE-89: SQL Injection via str.format() into execute()
    name = request.args.get("name")
    cur = db.cursor()
    cur.execute("SELECT * FROM users WHERE name = '{0}'".format(name))
    return str(cur.fetchall())


@app.route("/sqli/percent", methods=["POST"])
def sqli_percent():
    # CWE-89: SQL Injection via %-formatting into execute()
    role = request.form["role"]
    cur = db.cursor()
    cur.execute("SELECT * FROM users WHERE role = '%s'" % role)
    return str(cur.fetchall())


@app.route("/cmd/system")
def cmd_system():
    # CWE-78: OS Command Injection via os.system with concatenation
    host = request.args.get("host")
    os.system("ping -c 1 " + host)
    return "pinged"


@app.route("/cmd/popen")
def cmd_popen():
    # CWE-78: OS Command Injection via subprocess.Popen(shell=True)
    target = request.args.get("target")
    proc = subprocess.Popen("nslookup " + target, shell=True, stdout=subprocess.PIPE)
    return proc.stdout.read()


@app.route("/cmd/ospopen")
def cmd_ospopen():
    # CWE-78: OS Command Injection via os.popen
    f = request.args.get("file")
    return os.popen("cat " + f).read()


@app.route("/code/eval")
def code_eval():
    # CWE-95: Code Injection via eval() on user input
    expr = request.args.get("expr")
    return str(eval(expr))


@app.route("/code/exec", methods=["POST"])
def code_exec():
    # CWE-95: Code Injection via exec() on user input
    body = request.form["code"]
    exec(body)
    return "executed"


@app.route("/deser/pickle", methods=["POST"])
def deser_pickle():
    # CWE-502: Deserialization of Untrusted Data via pickle.loads
    raw = request.get_data()
    obj = pickle.loads(raw)
    return str(obj)


@app.route("/deser/yaml", methods=["POST"])
def deser_yaml():
    # CWE-502: Unsafe YAML deserialization (yaml.Loader allows object construction)
    data = request.form["doc"]
    obj = yaml.load(data, Loader=yaml.Loader)
    return str(obj)


@app.route("/import")
def dynamic_import():
    # CWE-470: Unsafe Reflection via __import__ on user-controlled module name
    mod = request.args.get("module")
    return str(__import__(mod))


@app.route("/header/sqli")
def header_sqli():
    # CWE-89: SQL Injection sourced from an HTTP header
    api_user = request.headers.get("X-User")
    cur = db.cursor()
    cur.execute(f"SELECT token FROM sessions WHERE user = '{api_user}'")
    return str(cur.fetchall())


@api.get("/api/search")
def fastapi_search(q: str = Query(...)):
    # CWE-89: SQL Injection from a FastAPI Query() parameter
    cur = db.cursor()
    cur.execute("SELECT * FROM products WHERE title LIKE '%" + q + "%'")
    return cur.fetchall()


@api.get("/api/run")
def fastapi_run(cmd: str = Query(...)):
    # CWE-78: OS Command Injection from a FastAPI Query() parameter
    return os.popen("sh -c " + cmd).read()


def cli_eval():
    # CWE-95: Code Injection via eval() on a CLI argument
    return eval(sys.argv[1])


def env_command():
    # CWE-78: OS Command Injection via environment variable into os.system
    payload = os.environ.get("USER_CMD", "")
    os.system("echo " + payload)


def stdin_import():
    # CWE-470: Unsafe Reflection via __import__ on a module name read from stdin
    line = sys.stdin.readline().strip()
    return __import__(line)


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=True)
